Implement and document safe operations on incomplete copies of BS4 elements

sknebel opened the pull request:
This modifies the modifying parts of backcompat.py to only affect the copied Tags, not the original ones, and I added a comment with some hints on what to do safely.

This still needs test ensuring it works: I'm thinking of changing all backcompat tests to create their own BS soups and additionally checking it for changes afterwards?

sknebel committed:
Implement and document safe operations on incomplete copies of BS4 elements

kartikprabhu commented:
If this works, I would recommend testing it in one test instead of adding tests to all the backcompat tests. Might be better for future organisation.

sknebel committed:
Add tests for not-modifying the tree (1 failing)

Tests that for all example HTML documents serializing the tree before and after parsing results in the same string.
Currently fails for <template> tags, since those still get deleted from the tree.

sknebel commented:
I had forgotten about the test generator pattern, that fits nicely. Added and surfaced one issue: I missed the deletion of `<template>` tags in `temp_fixes.py`.

sknebel commented:
3 options I see right now: 
a) jump over template tags while iterating, but this would let `<template>` tags leak through into `e-*` properties
b) replace template tags at start of processing with empty placeholders, exchange back after processing (or locally for only those in `e-*`, modifying the tree but repairing it afterwards (which would be a risk with multithreaded accesses to the same tree at the same time)
c) copying the subtrees for `e-*` properties or `e-*` properties containing `<template>` tags. 

b)/c) seem annoyingly much work and potential performance impact for such a special case, but a) might cause issues with downstream consumers not understanding `<template>` tags – opinions on if the spec prescribes a particular behavior here are split.

sknebel committed:
Fix handling of <template> tags while reducing copying

This changes the processing of template tags: <template> tags are skipped during normal parsing, and when an e-* property is parsed the template tag is removed from the subtree prior to turning it into HTML.
If a BeautifulSoup object was passed in originally, this happens on a copy of the subtree, leaving the original intact.

sknebel commented:
Template tags are fixed in this last commit, I'd consider this patch series finished with this and think it'd be ready for a new release of mf2py.

snarfed commented:
this looks great! thank you @sknebel, nice work. i'm going to merge this here and into upstream now, add `lxml` to the test requirements, and then announce the release candidate.

sknebel commented:
oh, lxml wasn't already? my mistake. we don't strictly need it there, so you could let mf2py pick a parser there I think.
